deployer/src/lib.rs

```diff
@@ -411,96 +437,78 @@ impl<D: Dal + Sync + 'static> Deployer for DeployerService<D> {
         Ok(Response::new(DestroyDeploymentResponse {}))
     }
 
+    type SubscribeProjectsStream =
+        tokio_stream::wrappers::ReceiverStream<TonicResult<ProjectEvent, tonic::Status>>;
+
     #[instrument(skip_all)]
-    async fn target_ip(
+    async fn subscribe_projects(
         &self,
-        request: tonic::Request<TargetIpRequest>,
-    ) -> TonicResult<tonic::Response<TargetIpResponse>, tonic::Status> {
-        // Authorize the request.
-        request.verify(Scope::ServiceRead)?;
-        let claim = request.extensions().get::<Claim>().cloned();
-        let request = request.into_inner();
-
-        // Do a cleanup in terms of previous invalid deployments.
-        let service_id = Ulid::from_string(&request.service_id)
-            .map_err(|_| tonic::Status::invalid_argument("invalid deployment id"))?;
-        let service = self
-            .dal
-            .service(&service_id)
-            .await
-            .map_err(|err| tonic::Status::not_found(err.to_string()))?;
-
-        if !service.state.is_running() {
-            return Err(tonic::Status::cancelled("the service is not running"));
-        }
-
-        if service.state.is_stopped() && request.instate {
-            let container = service.state.container().ok_or(tonic::Status::internal("the service is an unknown state, it's stopped but it doesn't have a container inspect info attached".to_string()))?;
-            let deployment_id = container
-                .deployment_id()
-                .map_err(|err| tonic::Status::internal(err.to_string()))?;
-            let service_name = container
-                .service_name()
-                .map_err(|err| tonic::Status::internal(err.to_string()))?;
-            let deployment =
-                self.dal
-                    .deployment(&deployment_id)
+        request: tonic::Request<SubscribeProjectsRequest>,
+    ) -> TonicResult<tonic::Response<Self::SubscribeProjectsStream>, tonic::Status> {
+        // request.verify(Scope::ProjectRead)?;
```

Comment:
This comes from gateway, so not sure which claim it should have :thinking:

Reply:
Exactly. I thought about an `Admin` scope but I am not yet clear how it will be able to get the JWT for it.